### PR TITLE
Fix run with progress and add new Run function

### PR DESCRIPTION
**Pull request by elivlo**
Hi,

I fixed the function RunWithProgress because there was a memory leak (line 212) when cancelling the context before the scan was finished.
I also added a new "Run" type of funtion (RunWithStreamer) that writes the xml-output directly to a file.

- This will prevent huge memory usage when running long scans.
- It also enables the standard nmap output to monitor the current nmap status.
- When the scan is finished the file can be parsed.

There is also an example showing how to run this type of scan.

Best regards!

P.S.
I would also appreciate a new tag :-)

### CODE REVIEW

File: examples/basic_scan_streamer_interface/main.go
Comment (Ullaakut) [inline examples/basic_scan_streamer_interface/main.go:19]:
```suggestion
// Write is a function that handles the normal nmap stdout.
```

File: examples/basic_scan_streamer_interface/main.go
Comment (Ullaakut) [inline examples/basic_scan_streamer_interface/main.go:21]: This variable is not used in this example, WDYT of just returning `nil` at the end?

File: examples/basic_scan_streamer_interface/main.go
Comment (Ullaakut) [inline examples/basic_scan_streamer_interface/main.go:30]:
```suggestion
// Bytes returns scan result bytes.
```

File: nmap.go
Comment (Ullaakut) [inline nmap.go:23]:
```suggestion
// Streamer constantly streams the stdout.
```

File: nmap.go
Comment (Ullaakut) [inline nmap.go:222]:
```suggestion
		// Trigger progress function exit.
```

File: nmap.go
Comment (Ullaakut) [inline nmap.go:282]:
```suggestion
// RunWithStreamer runs nmap synchronously. The XML output is written directly to a file.
```

File: nmap.go
Comment (Ullaakut) [inline nmap.go:284]: WDYT of removing the error here, since this func doesn't need error handling?

```suggestion
func (s *Scanner) RunWithStreamer(stream Streamer, file string) (warnings []string) {
```

File: nmap_test.go
Comment (Ullaakut) [inline nmap_test.go:273]: Could you please add a unit test for the new method? 🙏

**Review by Ullaakut — CHANGES_REQUESTED** [review]
Hi @elivlo ! Thanks for your continued interest in the project :) I haven't had much time to work on it these days so I really appreciate your help. Would you like to become maintainer of the project?

I have some minor suggestions for this specific PR but the idea seems good to me. I did not test it yet though.

File: nmap_test.go
Comment (elivlo) [reply nmap_test.go:273]: I will do so :)

File: nmap.go
Comment (elivlo) [reply nmap.go:284]: I think that is a good question. The error will show that the starting process could not launched. The warnings instead are stdout errors that do not mean a failed run.

- Instead of writing crucial errors to warnings I could return an error (l. 320-325)

WDYT of the proposed change?

**Comment by elivlo** [comment]
> Hi @elivlo ! Thanks for your continued interest in the project :) I haven't had much time to work on it these days so I really appreciate your help. Would you like to become maintainer of the project?
> 
> I have some minor suggestions for this specific PR but the idea seems good to me. I did not test it yet though.

Hi, let me say it is an honour for me to have the opportunity to become a maintainer! I would really appreciate it! But I also cannot say how much time I'm able to spend for maintaining. I will try my best.

File: nmap.go
Comment (Ullaakut) [reply nmap.go:284]: Sounds good to me

**Comment by Ullaakut** [comment]
@elivlo I've sent you an invite, and no worries, no strings attached :) This way though, you can even create the tag directly once this gets merged 👍

File: nmap.go
Comment (elivlo) [reply nmap.go:274]: I changed these lines :)

File: nmap_test.go
Comment (elivlo) [reply nmap_test.go:287]: I have written a unit test for this method but sadly I can't test all the errors until the code is refactored.

**Comment by elivlo** [comment]
@Ullaakut Hi, I have resolved your suggestions.

- I also added a few lines in the .travis.yml script. I hope this resolves the CI errors.

File: nmap.go
Comment (Ullaakut) [inline nmap.go:15]:
```suggestion
	"io"
	"os/exec"
	"strings"
	"time"
	
	"github.com/pkg/errors"
	"golang.org/x/sync/errgroup"
```

File: nmap.go
Comment (Ullaakut) [inline nmap.go:325]:
```suggestion
	// Check for warnings that will inevitably lead to parsing errors, hence, have priority.
```

File: nmap_test.go
Comment (Ullaakut) [inline nmap_test.go:18]: Can be made private

```suggestion
type testStreamer struct {
```

File: nmap_test.go
Comment (Ullaakut) [inline nmap_test.go:2130]:
```suggestion
			warnings: []string{"NoWarning", "NoWarning"},
```

File: nmap_test.go
Comment (Ullaakut) [inline nmap_test.go:2147]:
```suggestion
}

```

**Review by Ullaakut — COMMENTED** [review]
A few more minor suggestions, but for me the test is fine as is even if we're not at 100% coverage :) Good job!

**Review by Ullaakut — APPROVED** [review]
LGTM, thanks for your contribution :)

**Pull request merged**
